pilot/pkg/networking/core/v1alpha3/cluster_builder.go

```diff
@@ -1143,15 +1142,24 @@ func (cb *ClusterBuilder) buildUpstreamClusterTLSContext(opts *buildClusterOpts,
 	return tlsContext, nil
 }
 
-func (cb *ClusterBuilder) setAutoSni(mc *MutableCluster, tls *networking.ClientTLSSettings) {
-	if mc != nil && features.EnableAutoSni && len(tls.Sni) == 0 {
-		if mc.httpProtocolOptions == nil {
-			mc.httpProtocolOptions = &http.HttpProtocolOptions{}
-		}
-		if mc.httpProtocolOptions.UpstreamHttpProtocolOptions == nil {
-			mc.httpProtocolOptions.UpstreamHttpProtocolOptions = &core.UpstreamHttpProtocolOptions{}
+// Set auto_sni if sni field is not explicitly set and if EnableAutoSni feature flag is enabled.
+// Set auto_san_validation if VerifyCertAtClient feature flag is enabled
+func (cb *ClusterBuilder) setAutoSniAndAutoSanValidation(mc *MutableCluster, tls *networking.ClientTLSSettings) {
+	if mc != nil && features.EnableAutoSni {
+		if len(tls.Sni) == 0 || (features.VerifyCertAtClient && len(tls.SubjectAltNames) == 0) {
+			if mc.httpProtocolOptions == nil {
+				mc.httpProtocolOptions = &http.HttpProtocolOptions{}
+			}
+			if mc.httpProtocolOptions.UpstreamHttpProtocolOptions == nil {
+				mc.httpProtocolOptions.UpstreamHttpProtocolOptions = &core.UpstreamHttpProtocolOptions{}
+			}
+			if len(tls.Sni) == 0 {
+				mc.httpProtocolOptions.UpstreamHttpProtocolOptions.AutoSni = true
```

Comment:
Don't we still want this even if VerifyCertAtClient is false?

Reply:
Yes, line 1149 is an OR condition so that we don't unnecessarily set empty attrs..did I miss something?